app/services/data_stage/process_school_changes.rb

```diff
@@ -0,0 +1,90 @@
+# frozen_string_literal: true
+
+module DataStage
+  class ProcessSchoolChanges < ::BaseService
+    include GiasTypes
+
+    UNHANDLED_ATTRIBUTES = %w[
+      administrative_district_code
+      administrative_district_name
+      school_type_code
+      school_type_name
+      section_41_approved
+      ukprn
+    ].freeze
+
+    def call
+      DataStage::SchoolChange.unhandled.status_changed.find_each do |change|
+        next if has_unhandled_attributes?(change.attribute_changes)
+
+        if change.attribute_changes.key? "school_status_code"
+          handle_status_code_change(change)
+        end
+      end
+    end
+
+  private
+
+    def has_unhandled_attributes?(attributes)
+      (UNHANDLED_ATTRIBUTES & attributes.keys).any?
+    end
+
+    def handle_status_code_change(change)
+      from_code = change.attribute_changes["school_status_code"]&.first
+      to_code = change.attribute_changes["school_status_code"]&.last
+
+      if open_status_code?(from_code) && !open_status_code?(to_code)
+        close_school(change)
+      elsif !open_status_code?(from_code) && open_status_code?(to_code)
+        open_school(change)
+      else
+        # just sync then
+        change.school.create_or_sync_counterpart!
+      end
+    end
+
+    def open_school(change)
+      school = change.school
+
+      # is there a predecessor
+      last_link = school.school_links.order(created_at: :asc).last
+
+      if last_link.nil? || last_link.link_type == "Predecessor"
+        DataStage::School.transaction do
+          # no predecessor or simple predecessor - add new school
+          school.create_or_sync_counterpart!
+
+          # transfer assets if predecessor
+          if last_link.present?
+            predecessor = ::School.find_by(urn: last_link.link_urn)
+            transfer_assets_from_predecessor(school: school.counterpart, predecessor: predecessor) if predecessor.present?
+          end
+
+          change.update!(handled: true)
+        end
+      end
+    end
```

Comment:
Can we handle `last_link` being `nil` in a separate branch from it being `predecessor`, rather than intertwining them?

Also, what's the difference between `DataStage::School.transaction` and `ActiveRecord::Base.transaction`? Should we be using the latter, given that asset changes deal with non-school records?

Reply:
I don't think there is any difference but I like to use the model that things are relating to, but you're right it should probably be `::School.transaction` as the assets are related to the live school not the staged one

Reply:
I want them to be wrapped in the transaction so that the `change.update!` is only done if the create/sync/transfer are done - not sure if we'd just be less DRY if I split this up?

Reply:
Moving this class to a separate PR